test/plugin/test_in_tcp.rb

```diff
@@ -255,4 +255,78 @@ def create_tcp_socket(host, port, &block)
       assert_equal 'hello', event[2]['msg']
     end
   end
+
+  sub_test_case "message_length_limit" do
+    data("batch_emit", { extract: "" }, keep: true)
+    data("single_emit", { extract: "<extract>\ntag_key tag\n</extract>\n" }, keep: true)
+    test "drop records exceeding limit" do |data|
+      message_length_limit = 10
+      d = create_driver(base_config + %!
+        message_length_limit #{message_length_limit}
+        <parse>
+          @type none
+        </parse>
+        #{data[:extract]}
+      !)
+      d.run(expect_records: 2, timeout: 10) do
+        create_tcp_socket('127.0.0.1', @port) do |sock|
+          sock.send("a" * message_length_limit + "\n", 0)
+          sock.send("b" * (message_length_limit + 1) + "\n", 0)
+          sock.send("c" * (message_length_limit - 1) + "\n", 0)
+        end
+      end
+
+      expected_records = [
+        "a" * message_length_limit,
+        "c" * (message_length_limit - 1)
+      ]
+      actual_records = d.events.collect do |event|
+        event[2]["message"]
+      end
+
+      assert_equal expected_records, actual_records
+    end
+
+    test "clear buffer, which results in making an incomplete record" do |data|
+      message_length_limit = 12
+      d = create_driver(base_config + %!
+        message_length_limit #{message_length_limit}
+        delimiter ";"
+        <parse>
+          @type json
+        </parse>
+        #{data[:extract]}
+      !)
+      d.run(expect_records: 1, timeout: 10) do
+        create_tcp_socket('127.0.0.1', @port) do |sock|
+          sock.send('{"message":', 0)
+          sock.send('"hello', 0)
+          sleep 1 # To make the server read data and clear the buffer here.
+          sock.send('world!"};', 0) # this will be incomplete and fail to parse
+          sock.send('{"k":"v"};', 0) # this will succeed to parse
+        end
+      end
+
+      logs = d.logs.collect do |log|
+        log.gsub(/\A\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} [-+]\d{4} /, "")
+      end
+      actual_records = d.events.collect do |event|
+        event[2]
+      end
+
+      assert_equal(
+        {
+          logs: [
+            '[info]: The buffer size exceeds \'message_length_limit\', cleared: limit=12 size=17 head="{\"message\":\"hello"' + "\n",
+            '[warn]: pattern not matched message="world!\"}"' + "\n"
```

Comment:
Trailing data of dropped record should be obviously dropped as well.
How about like this:

```diff
diff --git a/lib/fluent/plugin/in_tcp.rb b/lib/fluent/plugin/in_tcp.rb
index 03e30d30..347458dc 100644
--- a/lib/fluent/plugin/in_tcp.rb
+++ b/lib/fluent/plugin/in_tcp.rb
@@ -106,6 +106,8 @@ module Fluent::Plugin
 
       log.info "listening tcp socket", bind: @bind, port: @port
       del_size = @delimiter.length
+      discard_till_next_delimiter = false
+
       if @_extract_enabled && @_extract_tag_key
         server_create(:in_tcp_server_single_emit, @port, bind: @bind, resolve_name: !!@source_hostname_key, send_keepalive_packet: @send_keepalive_packet) do |data, conn|
           unless check_client(conn)
@@ -120,6 +122,11 @@ module Fluent::Plugin
             msg = buf[pos...i]
             pos = i + del_size
 
+            if discard_till_next_delimiter
+              discard_till_next_delimiter = false
+              next
+            end
+
             if !@message_length_limit.nil? && @message_length_limit < msg.bytesize
               log.info "The received data is larger than 'message_length_limit', dropped:", limit: @message_length_limit, size: msg.bytesize, head: msg[...32]
               next
@@ -146,6 +153,7 @@ module Fluent::Plugin
           if !@message_length_limit.nil? && @message_length_limit < buf.bytesize
             log.info "The buffer size exceeds 'message_length_limit', cleared:", limit: @message_length_limit, size: buf.bytesize, head: buf[...32]
             buf.clear
+            discard_till_next_delimiter = true
             next
           end
         end
@@ -164,6 +172,11 @@ module Fluent::Plugin
             msg = buf[pos...i]
             pos = i + del_size
 
+            if discard_till_next_delimiter
+              discard_till_next_delimiter = false
+              next
+            end
+
             if !@message_length_limit.nil? && @message_length_limit < msg.bytesize
               log.info "The received data is larger than 'message_length_limit', dropped:", limit: @message_length_limit, size: msg.bytesize, head: msg[...32]
               next
@@ -189,6 +202,7 @@ module Fluent::Plugin
           if !@message_length_limit.nil? && @message_length_limit < buf.bytesize
             log.info "The buffer size exceeds 'message_length_limit', cleared:", limit: @message_length_limit, size: buf.bytesize, head: buf[...32]
             buf.clear
+            discard_till_next_delimiter = true
             next
           end
         end
```

Reply:
I see! I will try to fix this in that direction!
`in_udp` on Non-Windows has this specification too, but it does not seem as easy as this fix.
So, certainly, it would be best to fix it for `in_tcp` so far now.

Reply:
Fixed as https://github.com/fluent/fluentd/compare/283a3b3db39a625782d311671412d81b8d7e7aa0..0d9c3209de4bf1feaff23be048a0d6818b4e7b98

Reply:
I will re-test and update https://github.com/fluent/fluentd/pull/4137#issuecomment-1501517920

Reply:
Please let me know if you O.K. to merge.

Reply:
> I will re-test and update [#4137 (comment)](https://github.com/fluent/fluentd/pull/4137#issuecomment-1501517920)

Done!

> Please let me know if you are O.K. to merge.

I'm OK to merge!

Reply:
Thanks!